packages/flutter_tools/lib/src/commands/create.dart

```diff
@@ -90,6 +128,48 @@ class CreateCommand extends FlutterCommand {
   @override
   String get invocation => '${runner.executableName} $name <output directory>';
 
+  // If it has a .metadata file with the project_type in it, use that.
+  // If it has an android dir and an android/app dir, it's a legacy app
+  // If it has an android dir and an android/src dir, it's a plugin
+  // If it has .ios and/or .android dirs, it's an application (nee module)
+  // If it has an ios dir and an ios/Classes dir, it's a plugin
+  // If it has no ios dir, no android dir, and no .ios or .android, then it's a package.
+  ProjectType _determineTemplateType(Directory projectDir) {
+    yaml.YamlMap loadMetadata(Directory projectDir) {
+      if (!projectDir.existsSync())
+        return null;
+      final File metadataFile =fs.file(fs.path.join(projectDir.absolute.path, '.metadata'));
+      if (!metadataFile.existsSync())
+        return null;
+      return yaml.loadYaml(metadataFile.readAsStringSync());
+    }
+
+    // If it exists, the project type in the metadata is definitive.
+    final yaml.YamlMap metadata = loadMetadata(projectDir);
+    if (metadata != null && metadata['project_type'] != null) {
+      return _stringToProjectType(metadata['project_type']);
+    }
+
+    // There either wasn't any metadata, or it didn't contain the project type,
+    // so try and figure out what type of project it is from the existing
+    // directory structure.
+    if (fs.directory(fs.path.join(projectDir.absolute.path, 'android', 'app')).existsSync())
```

Comment:
Slightly more concise way to express this:

```dart
projectDir.absolute.childDirectory(fs.path.join('android', 'app'))
```

(and a few below as well if you feel like changing it)